src/ngx_http_lua_upstream_module.c

```diff
@@ -98,10 +84,854 @@ ngx_http_lua_upstream_create_module(lua_State * L)
+    lua_setfield(L, -2, "remove_peer");
+
+    return 1;
+}
+
+
+/*
+ * The function is compare server as upstream server 
+ * if exists and return upstream_server_t else return 
+ * NULL
+ * 
+*/
+static ngx_http_upstream_server_t*
+ngx_http_lua_upstream_compare_server(ngx_http_upstream_srv_conf_t * us , ngx_url_t u )
+{
+    ngx_uint_t                       i,j;
+    size_t                           len;
+    ngx_http_upstream_server_t      *server = NULL;
+
+    if (us->servers == NULL || us->servers->nelts == 0) {
+        return NULL;
+    }
+
+    server = us->servers->elts;
+
+    for (i = 0; i < us->servers->nelts; ++i) {
+        for(j = 0; j < server[i].naddrs; ++j) {
+
+            len = server[i].addrs[j].name.len;
+            if ( len == u.url.len 
+                 && ngx_memcmp( u.url.data ,server[i].addrs[j].name.data , u.url.len) == 0 ) {
+
+                 return  &server[i];
+            } 
+         }
+        
+    }
+
+    return NULL;
+
+}
+
+
+/*
+ * The function is dynamically add a server to upstream 
+ * server ,but not added it to the back-end peer
+ * 
+*/
+static int
+ngx_http_lua_upstream_add_server(lua_State * L)
+{
+
+    ngx_str_t                         host;
+    ngx_http_upstream_server_t       *us;
+    ngx_http_upstream_srv_conf_t     *uscf;
+    ngx_url_t                         u;
+    ngx_http_request_t               *r;
+    ngx_int_t                         weight, max_fails;
+    time_t                            fail_timeout;
+    u_char                           *p;
+
+    if (lua_gettop(L) != 5) {
+        // five param is :"upstream name", "ip:port" , "weight" , "max_fails", 
+        //"fail_time"
+        // for lua code , you must pass this five param, is none ,you should 
+        // consider pass default value.
+        return luaL_error(L, "exactly five argument expected");
+    }
+
+    r = ngx_http_lua_get_request(L);
+    if (NULL == r) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "get request error \n");
+        return 2;
+    }
+
+    host.data = (u_char *) luaL_checklstring(L, 1, &host.len);
+
+    ngx_memzero(&u, sizeof (ngx_url_t));
+    p = (u_char *) luaL_checklstring(L, 2, &u.url.len);
+    u.default_port = 80;
+
+    weight = (ngx_int_t) luaL_checkint(L, 3);
+    max_fails = (ngx_int_t) luaL_checkint(L, 4);
+    fail_timeout = (time_t) luaL_checklong(L, 5);
+#if (NGX_DEBUG)
+    ngx_log_error(NGX_LOG_ALERT, r->connection->log, 0, "%s %s params: %s,%s,%d,%d,%d\n", __FILE__,__FUNCTION__, host.data, p, weight, max_fails, fail_timeout);
+#endif
+    uscf = ngx_http_lua_upstream_find_upstream(L, &host);
+    if (NULL == uscf) {
+        lua_pushnil(L);
+        lua_pushliteral(L, "upstream not found\n");
+        return 2;
+    }
+ 
+     
+    // lua virtual machine memory is stack,so dup a memory 
+    u.url.data = ngx_pcalloc(uscf->servers->pool,u.url.len+1);
+    ngx_memcpy(u.url.data ,p ,u.url.len);
+
+    if ( NULL != ngx_http_lua_upstream_compare_server(uscf,u) ) {
```

Comment:
Style: please follow the nginx coding style. It requires the following form instead:

```C
    if (ngx_http_lua_upstream_compare_server(uscf,u) != NULL) {
```

No parentheses after `(` nor before `)`. And `!= NULL` should be on the right hand side.